content.js

```diff
@@ -0,0 +1,79 @@
+/*
+ * Referer Modifier: Modify the Referer header in HTTP requests
+ * Copyright (C) 2017-2020 Fiona Klute
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ */
+"use strict";
+
+// [NUANCE]
+//  It is undefined whether static content scripts will run before
+//  or after dynamic content scripts. If the dynamic script ran first,
+//  engineConfig is prefilled here, otherwise engineInstance.setConfig()
+//  will be called later.
+var engineConfig;
+var engineInstance = new RefererModEngine(engineConfig);
+
+(function() {
+
+	const originalGetter = Reflect.getOwnPropertyDescriptor(Document.wrappedJSObject.prototype, "referrer").get;
+
+	const documentMap = new WeakMap();
+
+	const _toString = Object.prototype.toString;
+	const _call = Function.prototype.call;
+
+	const dummy = {
+		get referrer() {
+			// `this` is an XPCNativeWrapper instance
+
+			// In case someone calls us on some random things
+			if (_toString.call(this) !== "[object HTMLDocument]" ||
+				_toString.call(Reflect.getPrototypeOf(this)) !== "[object HTMLDocument]"
+			) {
+				return _call.call(originalGetter, this);
+			}
```

Comment:
In what kind of situation could this happen?

Reply:
```
var foo = Reflect.getOwnPropertyDescriptor(Document.prototype, 'referrer').get.call({});
var bar = Document.prototype.referrer;
```

Reply:
Okay, I can see the issue. So the "damage" is only that someone could detect the extension, but still good to avoid. :sweat_smile:

Why not use
```js
if (!(this instanceof HTMLDocument))
```
for the check, though?

Reply:
Call me paranoid, but I was accounting for the case where `this` is from other window/document, such as `<iframe>`, `dispatchEvent`.
Admittedly, I have not tested everything thoroughly so I err on the side of paranoid.

Reply:
As for why I don't trust `instanceof`? `Object.create(HTMLDocument.prototype) instanceof HTMLDocument`.
While we are protected by `XPCNativeWrapper` and this might not apply here, I think `instanceof` does too many "magic" for my taste.

Reply:
Okay, I see your point about `instanceof`. :grin: 

While reading up on this I unfortunately found a workaround for the current approach, too: :sweat_smile: 

```js
function C() {}
C.prototype[Symbol.toStringTag] = 'HTMLDocument';
var c = new C();
Object.prototype.toString.call(c);
Object.prototype.toString.call(Reflect.getPrototypeOf(c));
```

If there isn't any reliable way to check the type and we can't (reliably) prevent detection that the `document.referrer` has been modified, I think the simple check

```js
if (Object.prototype.toString.call(this) !== "[object HTMLDocument]")
```

should be good enough to avoid messing anything up accidentally.

Reply:
> While reading up on this I unfortunately found a workaround for the current approach, too:
> ...

Got "[object Object]" in content script.
Cannot fool content scripts due to `XPCNativeWrapper`. See https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/Sharing_objects_with_page_scripts#xray_vision_in_firefox

Firefox ROCKS! This will fool content scripts in Chromium.

> `if (Object.prototype.toString.call(this) !== "[object HTMLDocument]")`
> should be good enough to avoid messing anything up accidentally.

Won't catch `HTMLDocument.prototype` then.

Reply:
You may also be interested in https://developer.mozilla.org/en-US/docs/Mozilla/Gecko/Script_security for more information on Gecko's security designs.

Reply:
I just confirmed that with some modification to the test environment, thank you! I had tested on the debug tools console before, and there the trick works. :sweat_smile:

The "Xray vision" and "Script security" pages are marked as archived on MDN, so I'm not sure how far we can still rely on them.

Reply:
> The "Xray vision" and "Script security" pages are marked as archived on MDN, so I'm not sure how far we can still rely on them.

As far as I can see (and tested), still all accurate. And I don't think Mozilla can easily replace those mechanisms, they are assumptions made implicitly by half the browser code. Mozilla will probably have to rewrite Firefox to invalidate those pages.